chia/wallet/puzzles/cat.clvm

```diff
@@ -0,0 +1,407 @@
+            inner_conditions
+          )
+      )
+    )
+
+     ;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+     (defun stager_three (
+         Truths
+         inner_conditions
+         prev_coin_id
+         this_coin_bundle
+         next_coin_id
+         prev_subtotal
+         extra_delta
+      )
+      (is_bundle_valid
+            this_coin_bundle
+            (limitations_program_from_cat_struct (cat_struct_truth Truths))
+            Truths
+            extra_delta
+            (generate_final_output_conditions
+              prev_subtotal
+              ; the expression on the next line calculates `this_subtotal` by adding the delta to `prev_subtotal`
+              (+ (+ prev_subtotal (- (input_amount_for_coin (coin_info_for_coin_bundle this_coin_bundle)) (output_totals inner_conditions))) extra_delta)
+              (morph_conditions inner_conditions (cat_struct_truth Truths))
+              prev_coin_id
+              (my_id_cat_truth Truths)
+              next_coin_id
+              (cat_struct_truth Truths)
+            )
+        )
+    )
+
+    (defun stager_two (
+        Truths
+        INNER_PUZZLE
+        inner_solution
+        prev_coin_id
+        this_coin_bundle
+        next_coin_id
+        prev_subtotal
+        extra_delta
+     )
+      (c (list ASSERT_MY_COIN_ID (my_id_cat_truth Truths))
+         (stager_three Truths
+                       (a INNER_PUZZLE (c Truths inner_solution))
+                       prev_coin_id
+                       this_coin_bundle
+                       next_coin_id
+                       prev_subtotal
+                       extra_delta
+         )
+
+      )
+   )
+
+
+
+    ; CAT TRUTHS struct is: ; CAT Truths is: (((innerpuzhash . my_amount) . (lineage_proof . cat_struct)) . ((my_id . full_puzhash) . (my_parent . ())))
+    ; create truths - this_coin_bundle verified true because we calculated my ID from it!
+    ; lineaege proof is verified later by cat parent check or genesis_checker_program
+
+    (defun stager (
+        cat_mod_struct
+        INNER_PUZZLE
+        inner_solution
+        inner_puzzle_hash
+        my_id
+        prev_coin_id
+        this_coin_bundle
+        next_coin_bundle
+        prev_subtotal
+        extra_delta
+     )
+      ; struct construction params - (my_id my_parent full_puzhash my_amount innerpuzhash lineage_proof cat_struct)
+      (stager_two (cat_truth_data_to_truth_struct my_id (f (f this_coin_bundle)) (f (r (f this_coin_bundle))) (f (r (r (f this_coin_bundle)))) inner_puzzle_hash (r this_coin_bundle) cat_mod_struct)
+             INNER_PUZZLE
+             inner_solution
+             prev_coin_id
+             this_coin_bundle
+             (coin_id_for_coin next_coin_bundle cat_mod_struct)
+             prev_subtotal
+             extra_delta
+       )
+    )
+
+    (stager
+        ;; caclculate cat_mod_struct, inner_puzzle_hash, coin_id
+        (list MOD_HASH (sha256 1 MOD_HASH) limitations_program LIMITATIONS_PROGRAM_HASH)
```

Comment:
It's very alarming to see `LIMITATIONS_PROGRAM_HASH` and `limitations_program` batched up together here without verifying that they match. It does seem to be checked later, but is there a good reason to not do it here?

Reply:
I see, the program is actually optional. In that case, consider renaming `limitations_program` to `optional_limitations_program` to increase clarity (and reserving `limitations_program` to the case where it's already been check as null and validated to match the hash).